locale/localization.js

```diff
@@ -0,0 +1,31 @@
+
+module.exports = {
+	addTranslationFunc : function (localeData){
+		// Add translation function to String prototype
+		String.prototype.translate = function(_groups = String.prototype.getTranslationDefaults()) {
+			const text = this.valueOf();
+			const groups = _groups ? Array.from(_groups) : [];
+			let obj = localeData;
+			while (groups?.length > 0) {
+				if(obj[groups[0]]) {
+					obj = obj[groups[0]];
+					groups.shift();
+					continue;
+				}
+				break;
+			}
+
+			if(obj[text] && obj[text].length > 0) return `${obj[text]}`;
+			return localeData[text]?.length > 0 ? `${localeData[text]}` : `${text}`;
+		};
```

Comment:
I would suggest we add a parameter **interpolations** to the `translate` function which is an object containing key-value pairs of any text that need to be interpolated in the text.

'UsersPublishedBrews`.translate({username : this.props.username});`

Then in the YAML file, we can just add some kind of marker (I would vote the standard `${}`):

`UsersPublishedBrews : '${username}'s published brews'`
...
`UsersPublishedBrews : 'Brews publicados de ${username}'`

```suggestion
String.prototype.translate = function(_groups = String.prototype.getTranslationDefaults(), interpolations = {}) {
    const text = this.valueOf();
    const groups = _groups ? Array.from(_groups) : [];
    let obj = localeData;
    while (groups?.length > 0) {
        if(obj[groups[0]]) {
            obj = obj[groups[0]];
            groups.shift();
            continue;
        }
        break;
    }

    let translatedText = obj[text] && obj[text].length > 0 ? `${obj[text]}` : localeData[text]?.length > 0 ? `${localeData[text]}` : `${text}`;

    // Interpolate the values
    for (let key in interpolations) {
        translatedText = translatedText.replace(new RegExp(`\${${key}}`, 'g'), interpolations[key]);
    }

    return translatedText;
};
```

Reply:
> Then in the YAML file, we can just add some kind of marker (I would vote the standard ${}):

Why partially emulate template strings when you could implement actual template strings?

```javascript
 if (translationData[key]) {
    // Dynamically define the function
    const functionString = `with (values) return \`${translationData[key]}\`;`;
    const translationFunction = new Function('name', functionString);

    // Call the dynamically defined function
    return translationFunction(values);
  } else {
    return `Translation not found for key: ${key}`;
  }
```

Reply:
Not opposed to that, but I don't see what the added benefit would be given the increased complexity. (The proposed code uses deprecated and highly discouraged syntax as well, so we should avoid that if going this route.)

Reply:
The proposal supports `` `${expression}` ``, which is more than simple substitution, and thus supports `` `${name.endsWith('s') ? name : name + 's'} Published Brews` `` (for example).

So .. odd grammar rules (like English possessive `'s`), or pluralisation rules, or even simply date formatting .. can all be supported via `` `${expression}` `` but not `` ${field}``.

The general concept of dynamically compiling a function from a string is usually discouraged because you really don't want to allow user-supplied strings being compiled. In this situation though only the strings in the translation files get compiled, and they should be getting reviewed when committed.

The other issue sometimes noted is that using `new Function` constructors can impair code optimisation and is less performant. Unless we want to hardcode a bunch of functions for use by `translate()`, we're gonna have to compile on the fly.

Reply:
The points about `new Function` are valid and something to be careful with. But my main concern is actually the `with` operator, which is deprecated and should not be used.

Reply:
I might suggest lodash's `_.template` function which can evaluate a string with expressions like this.

https://lodash.com/docs/#template

Reply:
Ah ..  my understanding with the `with` operator is that the main discouragment has to do with potentially causing unintended variable declarations and making it difficult to determine where variables are defined. Last I heard it wasn't actually deprecated, per se.

Nonetheless, I support the recommendation of `_.template` .. it has additional checks too (but still uses `new Function` as the fundamental technique, heh).

Reply:
So for reference, I believe the `_.template` could be used like this:

```
String.prototype.translate = function(_groups = String.prototype.getTranslationDefaults(), interpolations = {}) {
    const text = this.valueOf();
    const groups = _groups ? Array.from(_groups) : [];
    let obj = localeData;
    while (groups?.length > 0) {
        if(obj[groups[0]]) {
            obj = obj[groups[0]];
            groups.shift();
            continue;
        }
        break;
    }
    let translatedText = obj[text] && obj[text].length > 0 ? `${obj[text]}` : localeData[text]?.length > 0 ? `${localeData[text]}` : `${text}`;

    // Interpolate the values
    var compiledTemplate = _.template(translatedText);
    translatedText = compiledTemplate(interpolations);
    return translatedText;
};
```